lib/action_subscriber/message_retry.rb

```diff
@@ -0,0 +1,38 @@
+module ActionSubscriber
+  module MessageRetry
+    def self.redeliver_message_with_backoff(env)
+      ttl = calculate_ttl(env)
+      return unless ttl
+      with_exchange(env, ttl) do |exchange|
+        exchange.publish(env.encoded_payload,
+          :content_type => env.content_type,
+          :routing_key => env.routing_key,
+          :expiration => ttl
+        )
+      end
+    end
+
+    # Private Implementation
+    def self.calculate_ttl(env)
+      death = env.headers.fetch("x-death", [{"original-expiration" => "20"}]).first
+      next_ttl = death["original-expiration"].to_s.to_i * 5
+      return false if next_ttl > 86_400_000
+      next_ttl
+    end
+
+    def self.with_exchange(env, ttl)
+      retry_name = "#{env.exchange}_retry_#{ttl}"
```

Comment:
Checking into this a bit further it doesn't look like queues cost much in terms of CPU or RAM. In our production system we are using about 19MB per rabbitmq instance for queues. So if we 10x the number of queues in the system that would take up about 190MB of RAM.

Each connection / channel / consumer takes up a lot more resources than the queues themselves. And since these queues will just be used to hold messages while we wait for them to expire we won't actually have any channels or consumers connected to them.

<img width="940" alt="screen shot 2015-10-20 at 9 54 40 am" src="https://cloud.githubusercontent.com/assets/80008/10613050/cd4d3fa0-7710-11e5-8855-05ac4486f8e9.png">